Added HCS tutorial: Batch, anchor, and verify records with HCS (#358)
hashgraph/hedera-docs: hedera/tutorials/consensus/batch-anchor-verify-records-with-hcs.mdx

---
title: "Batch, anchor, and verify records with HCS"
description: "Learn how to batch records off-chain, compute a Merkle root, and anchor it on Hedera Consensus Service for cost-effective verification."
---

The Hedera Consensus Service (HCS) enables decentralized event ordering and immutable timestamping for any application. A best practice for data integrity involves anchoring a 'digital fingerprint' of your records on-chain, which provides a verifiable audit trail without exposing sensitive information. Merkle roots are cryptographic summaries that enable the efficient verification of large datasets, allowing you to also prove the existence of individual records within a batch. This tutorial demonstrates how to use these tools to verify data on a public ledger like Hedera in a manner that is both highly secure and cost-effective.

## What You Will Accomplish
- Compute a Merkle root from a batch of off-chain records
- Anchor that Merkle root on HCS using `ConsensusSubmitMessage`
- Verify the batch (and a single record) using the mirror node

## Prerequisites
- Node.js
- A Hedera testnet account (see [Hedera Portal](https://portal.hedera.com))

> ## Table of Contents

Reviewer: ```suggestion
## Table of Contents
1. [Setup and Installation](#1-setup-and-installation)
2. [Understand the dataset](#2-understand-the-dataset-you-will-anchor)
3. [Create a topic for batch anchoring and verification](#3-create-a-topic-for-batch-anchoring-and-verification)
4. [Compute the Merkle root](#4-compute-the-merkle-root-local)
5. [Anchor the Merkle root on HCS](#5-anchor-the-merkle-root-on-hcs)
6. [Verify the batch via Mirror Node](#6-verify-the-anchored-root-using-the-mirror-node)
7. [Verify a single record (Proof)](#7-verify-a-single-record-using-a-merkle-proof)
8. [Next steps](#next-steps)
```

1. Setup and Installation
2. Understand the dataset
3. Create a topic for batch anchoring and verification
4. Compute the Merkle root
5. Anchor the Merkle root on HCS
6. Verify the batch via Mirror Node
7. Verify a single record (Proof)
8. Next steps

---

## 1. Setup and Installation

### 1a. Clone and Install
Clone the repository and install dependencies:

```bash
git clone https://github.com/hedera-dev/tutorial-hcs-batching-hashing-verifying-js.git
cd tutorial-hcs-batching-hashing-verifying-js
npm install
```

### 1b. Configure Environment
Copy the example environment file:

```bash
cp .env.example .env
```

> Open `.env` and fill in your Testnet credentials:

Reviewer: ```suggestion
Open `.env` and fill in your Testnet credentials:
- `OPERATOR_ID`: Your Hedera account ID (e.g. `0.0.12345`)
- `OPERATOR_KEY`: Your **ECDSA DER-encoded private key** (for example, a key starting with `3030...`)
- `HEDERA_NETWORK`: `testnet`
- `MIRROR_NODE_BASE_URL`: Leave as `https://testnet.mirrornode.hedera.com`
```

  module.exports = {
      createTopic,
      submitMessage
  };
  ```
  </CodeGroup>
</Accordion>

**Expected Output:**
```
--- 3. Anchor Batch Merkle Root on HCS ---
...
1) Recomputed local Merkle Root: 1d59720e...
2) Built anchor message (215 bytes).

Submitting to Topic 0.0.98765...

✅ Message Anchored!
   Transaction ID: 0.0.1307@1767056727.814369284
   HashScan: https://hashscan.io/testnet/transaction/0.0.1307@1767056727.814369284
   Status: SUCCESS
   Merkle Root: 1d59720e...
```
This approach is efficient because instead of sending 100 individual transactions, you send **one** transaction with the Merkle root.

---

## 6. Verify the anchored root using the mirror node

With the Merkle root hash on the public ledger, anyone can verify the batch integrity. Running `scripts/04-verify-batch.js` confirms this by completing the following steps:
  1.  **Recompute Root**: Loads the local dataset and calculates the Merkle root from your local `data/batch-100.json` exactly as before using `computeRoot`.
  2.  **Fetch Message**: Queries the [Mirror Node REST API](https://docs.hedera.com/hedera/sdks-and-apis/rest-api) for the latest message on the topic using `getLatestTopicMessage`.

>   3.  **Compare**: decoding the message and verifying that the on-chain root matches the locally computed root.

Reviewer: ```suggestion
  3.  **Compare**: Decode the message and verify that the on-chain root matches the locally computed root.
```

  const { sha256 } = require('./hash');

  /**
    * Verifies a Merkle proof.
    * @param {string} leafHashHex 
    * @param {string} rootHashHex 
    * @param {Array} proof 
    * @returns {boolean}
    */
  function verifyProof(leafHashHex, rootHashHex, proof) {
      let currentHash = Buffer.from(leafHashHex, 'hex');

      for (const step of proof) {
          const siblingParams = Buffer.from(step.siblingHashHex, 'hex');

          if (step.position === 'right') {
              currentHash = sha256(Buffer.concat([currentHash, siblingParams]));
          } else {
              currentHash = sha256(Buffer.concat([siblingParams, currentHash]));
          }
      }

      return currentHash.toString('hex') === rootHashHex;
  }

  module.exports = {
      computeRoot,
      getProof,
      verifyProof
  };

  ```
  </CodeGroup>
</Accordion>

**Output:**
```
✅ PASS: Record "record-042" is cryptographically proven to be in the batch.
```

---

## Code Check ✅

It's time to try this example yourself! Get the code from the GitHub repository:
[tutorial-hcs-batching-hashing-verifying-js](https://github.com/hedera-dev/tutorial-hcs-batching-hashing-verifying-js)

> ## Notes on message limits

Reviewer: ```suggestion
## Notes on Message Limits
```
